### DISCUSSION TITLE
Self-referencing many-to-many field

### DISCUSSION THREAD

**Discussion by TopchetoEU** [mikro-orm/mikro-orm#6190] [Q&A]
Currently I'm trying to create something like a node graph of one of my models, something like this:

```ts
@Entity()
export class KidEntity {
    @PrimaryKey({ type: "uuid" })
    id!: string;
    @ManyToMany(() => KidEntity, v => v.siblings, { owner: true })
    siblings = new Collection<KidEntity>(this);
}
```

However, MikroORM doesn't seem to support such a feature, as it throws the following error:
```
Both KidEntity.siblings and KidEntity.siblings are defined as owning sides, use 'mappedBy' on one of them
```

So far a workaround seems to be to define two collections like this:

```ts
@Entity()
export class KidEntity {
    @PrimaryKey({ type: "uuid" })
    id!: string;
    @ManyToMany(() => KidEntity, v => v.siblings2, { owner: true })
    siblings1 = new Collection<KidEntity>(this);

    @ManyToMany(() => KidEntity, v => v.siblings1, { owner: false })
    siblings2 = new Collection<KidEntity>(this);
}
```

This however is a very dirty solution, as you need to work with both collections in parallel, which is error-prone and defeats the purpose of having an ORM (in my opinion).

A solution might be to either create a special case for the existing `m:m` relationship when the owning and inverse side are the same field, create a special flag for such cases in the options for the ManyToMany decorator or just create a new relationship entirely.

I've dug a little around in the code base to find a possible way to implement such a relationship myself, but it seems that relationship types are just hard-coded. This means that implementing this feature would mean essentially creating a new relationship type (a bidirectional graph, opposed to the usual unidirectional graph that is a `m:m` relationship), which would be a breaking change which would break the different database adapters (which could be very bad).

If my observations are true, I would understand hesitance to implement such a feature, but at least consider it for a future major release, as I believe this is a very important (albeit rarely used) feature for a database.

If such a feature exists tho, please document it.

**Comment by B4nan** [comment]
> So far a workaround seems to be to define two collections like this:

This is the solution, not a workaround. You can't define bidirectional relations with just one side. 

You can however define just the owning side.

```ts
@Entity()
export class KidEntity {
    @PrimaryKey({ type: "uuid" })
    id!: string;
    @ManyToMany(() => KidEntity)
    siblings = new Collection<KidEntity>(this);
}
```

I don't understand why you think this is "dirty" and error-prone. If you have just one side, how can you distinguish the directions? Relations are always directional, including M:N.

**Reply by B4nan** [reply]
It feels like you just want the ORM to fix your DB design. I don't think M:N is the right way to model siblings like in your example, you should have a parent relation and map it based on that instead. Siblings are those with the same parent, why would you need a pivot table for that?

**Reply by TopchetoEU** [reply]
My use case is more complex as a kid may have more than one parents. Also  a kid may have friends, and if Josh is a friend of Pete, than, of course, Pete is a friend of Josh. I haven't found a clean way to model this using this ORM. Admittedly, I should've used the friends m:m relationship as an example instead.

**Reply by TopchetoEU** [reply]
Also, to the suggestion for a unidirectional many-to-many to the same entity worked, but if I added Pete in John's siblings list, then John wouldn't be in Pete's sibling list. I believe there should be a way to model such a relationship (aka a bidirectional graph).

**Reply by B4nan** [reply]
> Admittedly, I should've used the friends m:m relationship as an example instead.

That example isn't any better, you might call me a friend while I don't call you a friend of mine :] That's still a directional relation to me.

**Reply by B4nan** [reply]
This all boils down to the fact that the ORM relations just represent your schema, and FKs are directional. I am not aware of any ORM that would support natively what you want, feel free to prove me wrong.

So either you will want to have two rows in that pivot table to support both directions (simpler for querying, harder for setting up), or just one (harder for querying, simpler for setting up).
